torch/csrc/distributed/autograd/context/thread_local_context.h

```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include <torch/csrc/distributed/autograd/context/context.h>
+
+namespace torch {
+namespace distributed {
+namespace autograd {
+
+class ThreadLocalDistAutogradContext {
+ public:
+  ThreadLocalDistAutogradContext();
+  explicit ThreadLocalDistAutogradContext(ContextPtr::weak_type&& context_wp);
+  ~ThreadLocalDistAutogradContext();
+
+  ContextPtr::weak_type context_weak_ptr_;
+
+ private:
+  ContextPtr::weak_type prev_context_weak_ptr_;
```

Comment:
What is the reason to use a weak_ptr here? I feel it's safer to pass in the shared_ptr here to ensure that we have a valid reference to the context as long as the guard is alive.

Reply:
My concern is a `thread_local shared_ptr` may unnecessarily extend the lifetime of the RPC context instance, because ` thread_local` has static storage.